src/celrender/gl/vertexobject.cpp

```diff
@@ -0,0 +1,273 @@
+    m_primitive = other.m_primitive;
+    m_count = other.m_count;
+    m_id = other.m_id;
+    m_idxBufferId = other.m_idxBufferId;
+    m_indexType = other.m_indexType;
+    m_currBuff = other.m_currBuff;
+    m_initialized = other.m_initialized;
+
+    other.m_primitive = Primitive::Triangles;
+    other.m_count = 0;
+    other.m_id = 0;
+    other.m_idxBufferId = 0;
+    other.m_indexType = IndexType::UnsignedShort;
+    other.m_currBuff = 0;
+    other.m_initialized = false;
+
+    return *this;
+}
+
+VertexObject::~VertexObject()
+{
+    clear();
+}
+
+VertexObject&
+VertexObject::setCount(int count)
+{
+    m_count = count;
+    return *this;
+}
+
+VertexObject&
+VertexObject::setPrimitive(VertexObject::Primitive primitive)
+{
+    m_primitive = primitive;
+    return *this;
+}
+
+void
+VertexObject::clear() noexcept
+{
+    if (m_id != 0 && isVAOSupported())
+        glDeleteVertexArrays(1, &m_id);
+    m_id = 0;
+}
+
+struct VertexObject::BufferDesc
+{
+    BufferDesc(
+        GLsizeiptr    offset,
+        GLuint        bufferId,
+        std::uint16_t type,
+        std::int16_t  location,
+        std::uint8_t  elemSize,
+        std::uint8_t  stride,
+        bool          normalized) :
+        offset(offset),
+        bufferId(bufferId),
+        type(type),
+        location(location),
+        elemSize(elemSize),
+        stride(stride),
+        normalized(normalized)
+    {
+    }
+    GLsizeiptr      offset;
+    GLuint          bufferId;
+    std::uint16_t   type;           // all constants < 0xFFFF
+    std::int16_t    location;
+    std::uint8_t    elemSize;       // 1, 2, 3, 4
+    std::uint8_t    stride;         // WebGL allows only 255 bytes max
+    bool            normalized;
+};
+
+VertexObject&
+VertexObject::addVertexBuffer(const Buffer &buffer, int location, int elemSize, VertexObject::DataType type, bool normalized, int stride, std::ptrdiff_t offset)
+{
+    if (buffer.targetHint() != Buffer::TargetHint::Array)
+        return *this;
+
+    m_bufferDesc.emplace_back(
+        offset,
+        buffer.id(),
+        static_cast<std::uint16_t>(type),
+        static_cast<std::uint16_t>(location),
+        static_cast<std::uint8_t>(elemSize),
+        static_cast<std::uint8_t>(stride),
+        normalized
+    );
+
+    return *this;
+}
+
+VertexObject&
+VertexObject::draw()
+{
+    return draw(m_primitive, m_count, 0);
+}
+
+VertexObject&
+VertexObject::draw(int count, int first)
+{
+    return draw(m_primitive, count, first);
+}
+
+VertexObject&
+VertexObject::draw(VertexObject::Primitive primitive, int count, int first)
+{
+    if (count == 0)
+        return *this;
+
+    current = this;
+
+    if (!m_initialized)
+    {
+        m_initialized = true;
+        if (isVAOSupported())
+        {
+            glBindVertexArray(m_id);
+            enableAttribArrays();
+            m_bufferDesc.clear();
+        }
+    }
+    else
+    {
+        if (isVAOSupported())
+            glBindVertexArray(m_id);
```

Comment:
I think we also need to bind the index buffer here in case of `isIndexed()` - at least this seems to resolve the segfault

Reply:
Do you have both in GNU/Linux and Windows?

Reply:
> I'm getting a segfault when the star labels are being rendered, need to get some time to investigate it.

Do you have it both in GNU/Linux and Windows? What HW do you use?

Reply:
Occurs on both Linux (Kubuntu 23.04) and Windows, using Nvidia GTX 1070 card. The error occurs in the `glDrawElements` call on line 207. Previous calls to the code for non-indexed buffers succeed, so does the first call to this method for an indexed buffer (which hits the `!m_initialized` code path). The second call, where the buffer is already initialized triggers the segfault.

Reply:
On both Windows and Linux, I can avoid the segfault by using:

```cpp
if (!m_initialized)
{
    // ... snip ...
}
else
{
    if (isVAOSupported())
    {
        glBindVertexArray(m_id);
        if (isIndexed())
            glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, m_idxBufferId);
    }
    else
        enableAttribArrays();
}
```

Reply:
Then this looks like a bug in Nvidia drivers. I'll add an workaround for them.

Reply:
please check now. (btw, there is still a bug in nvidia drivers because they should not segfault)